ta/sdp_basic/ta_sdp_basic.c

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <string.h>
+#include <tee_api.h>
+#include <tee_internal_api_extensions.h>
+#include <tee_internal_api.h>
+#include <tee_ta_api.h>
+#include <trace.h>
+
+#include <ta_sdp_basic.h>
+
+#define PTA_SELF_TEST_UUID \
+		{ 0xd96a5b40, 0xc3e5, 0x21e3, \
+			{ 0x87, 0x94, 0x10, 0x02, 0xa5, 0xd5, 0xc6, 0x1b } }
+
+#define PTA_SELF_TEST_CMD_INJECT_SDP		3
+#define PTA_SELF_TEST_CMD_TRANSFORM_SDP	4
+#define PTA_SELF_TEST_CMD_DUMP_SDP		5
+
+/*
+ * Basic Secure Data Path access test commands:
+ * - command INJECT: copy from non secure input into secure output.
+ * - command TRANSFROM: read, transform and write from/to secure in/out.
+ * - command DUMP: copy from secure input into non secure output.
+ */
+
+static TEE_Result cmd_inject(uint32_t types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result rc;
+	const int sec_idx = 1;		/* highlight secure buffer index */
+	const int ns_idx = 0;		/* highlight nonsecure buffer index */
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters %x", (unsigned)types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (params[sec_idx].memref.size < params[ns_idx].memref.size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	/* intentionally check the overall buffers permissions */
```

Comment:
Ok, i must be more explicit in this comment.
The fact is that here we could only check the secure attribute of the 1st byte of the buffer, it will safely reflect the attribute of the whole buffer. Checking the overall buffer is a waste of time but since this test does not care about performances, lets check the secure attribute for the overall buffer length.